### DISCUSSION TITLE
How do I get IMenuCommandService?

### DISCUSSION THREAD

**Discussion by LadyCailin** [VsixCommunity/Community.VisualStudio.Toolkit#349] [Q&A] [answered]
I am trying to modify the disabled state of various menu buttons in my extension based on dynamic factors. I think I see how to do this, but all the examples seem to require getting a reference to the IMenuCommandService, and every way I've tried getting it always returns null (or throws an exception, in the case of VS.GetRequiredService). I was following other examples, at least in code, but I might be missing something somewhere else, like in the manifest or something, I dunno. 

In all these cases, I've trying to call it in InitializeAsync of the ToolkitPackage, my thought is that maybe this is too early, but I have seen other examples (for instance [here](https://github.com/madskristensen/FileNesting/blob/0d043da031a173fe874ec7aa5eb138079c77fa43/src/VSPackage.cs)), though admittedly this is optional.

I am autoloading at the NoSolution level, though delaying that to SolutionExists and CodeWindow didn't seem to help either, so I don't know if that matters.

I can show code examples if I need to, but I was matching the above code from the FileNesting extension, and that doesn't work, so I must be missing something somewhere else that I don't know about.

**Comment by reduckted** [comment, marked as answer]
The `IMenuCommandService` can only be retrieved from a VSPackage or tool window (see https://github.com/VsixCommunity/Community.VisualStudio.Toolkit/issues/92#issuecomment-888543039).

So in your package class, instead of calling this:
```cs
commandService = await VS.GetServiceAsync<IMenuCommandService, IMenuCommandService>();
// ❌ Will throw an exception.
```

You instead need to call this:
```cs
commandService = await this.GetServiceAsync<IMenuCommandService, IMenuCommandService>();
```

**Reply by LadyCailin** [reply]
Aha! Yes, that was exactly it. I thought I had tried that already, but I tried so many things at this point that who knows what I might have been doing wrong. Anyways, that does work! However, I am now on to the next problem, which is that FindCommand returns null for commands which were created through the vsct file. Is this expected, since I didn't add the commands manually with AddCommand?

**Reply by reduckted** [reply]
Yes, that would be expected. Even if the command is defined in the `.vsct` file, you still need to add the command to the `IMenuCommandService`. The `.vsct` only describes the commands in a package. This allows Visual Studio to add menu items and key bindings without needing to load your extension. When a command is executed, Visual Studio will make sure your extension is loaded, and expects the extension to add the commands at that point.

**Reply by LadyCailin** [reply]
Hmm, that's a bit of a shame, since this is duplicated code basically. Seems like something the ToolkitPackage could automatically parse and add for me, though perhaps that has undesired side effects I'm not aware of. If that sounds like a good feature request to you though, I can file it.

**Reply by LadyCailin** [reply]
Oh, but actually, the handler for the command is automatically registered for me when I add the CommandAttribute and override ExecuteAsync. So I'm not sure what to add in the MenuCommand for the handler.

**Reply by reduckted** [reply]
Are you calling `await this.RegisterCommandsAsync();` in your package's `InitializeAsync` method? For example:

https://github.com/VsixCommunity/Community.VisualStudio.Toolkit/blob/9c32305c83827d0bfd7c85777f9e79fc71e22c2b/demo/VSSDK.TestExtension/TestExtensionPackage.cs#L34-L38

**Reply by LadyCailin** [reply]
Aha, yes, this is working now, thanks! I'll study this example too, I didn't know of its existence.
